dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java

```diff
@@ -57,12 +61,27 @@ public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWork
           "Not enough workers in the cluster %d workers in the cluster but %d required",
           blockWorkerInfos.size(), count));
     }
-    HASH_PROVIDER.refresh(blockWorkerInfos, mNumVirtualNodes);
-    List<BlockWorkerInfo> workers = HASH_PROVIDER.getMultiple(fileId, count);
+    List<WorkerIdentity> workerIdentities = blockWorkerInfos.stream()
+        .map(BlockWorkerInfo::getIdentity)
+        .collect(Collectors.toList());
+    mHashProvider.refresh(workerIdentities, mNumVirtualNodes);
+    List<WorkerIdentity> workers = mHashProvider.getMultiple(fileId, count);
     if (workers.size() != count) {
       throw new ResourceExhaustedException(String.format(
           "Found %d workers from the hash ring but %d required", workers.size(), count));
     }
-    return workers;
+    ImmutableList.Builder<BlockWorkerInfo> builder = ImmutableList.builder();
+    // todo(bowen): this is quadratic complexity. examine if it's worthwhile to replace
```

Comment:
or an extra `Map<WorkerIdentity, BlockWorkerInfo>` may give you `O(1)`

Reply:
I think we need a better design here.
The performance to compare such a large list is high.

If we have some version control/version number of this list, we can know if the list is changed since last time it is access.